draft-nygren-dnsop-svcb-httpssvc.md

```diff
@@ -659,24 +661,21 @@ The SvcParamKey for ESNI is "esnikeys".  Its value is defined in
 
 The "a" and "aaaa" keys represent IP address hints for the service.
 If A and AAAA records for SvcDomainName are locally available, the client SHOULD
-ignore these hints.  When A and AAAA records arrive, clients SHOULD
+ignore these hints.  Otherwise, clients MUST perform A and/or AAAA queries
+for SvcDomainName as in {{client-behavior}}, and clients SHOULD
 switch to an IP address in those records as soon as possible.
 
 The wire format for each parameter is a sequence of IP addresses in network
 byte order.  Like an A or AAAA RRSet, the list of addresses represents an
-unordered collection.
+unordered collection, and clients SHOULD pick addresses to use in a random order.
 
 These parameters MAY be repeated multiple times within a record.
-When this happens, the list of addresses SHOULD be concatenated.
+When receiving such a record, clients SHOULD combine the sets of addresses.
 
 When selecting between AAAA and A records to use, clients may use an
-approach such as {{!HappyEyeballsV2=RFC8305}}.  Within an address
-family, clients SHOULD pick addresses to use in a random order.
-
-When only "a" parameters are present, IPv6-only clients MUST either
-resolve an AAAA records from the SvcDomainName or attempt
-to synthesize IPv6 addresses as specified in {{!RFC7050}}.
-This is necessary as recursive resolvers will not be able to
+approach such as {{!HappyEyeballsV2=RFC8305}}.
+When only "a" parameters are present, IPv6-only clients may synthesize
+IPv6 addresses as specified in {{!RFC7050}}.  Recursive resolvers MUST NOT
```

Comment:
I'm still solidly in the "if you are skipping your local DNS resolver's DNS64, you MUST do rfc7050 synthesis" camp.  There are tens (hundreds?) of millions of mobile end-users who will be broken otherwise for IPv4-only sites.  An alternative would be to say that a "aaaa" parameter MUST be included if an "a" parameter is present?  (Which puts the onus on the server-side.)  If we don't have a MUST on the client or the server, recursive resolvers will do DNS64 synthesis here so having a MUST NOT there is unrealistic.

Reply:
> I'm still solidly in the "if you are skipping your local DNS resolver's DNS64, you MUST do rfc7050 synthesis" camp.

Me too.

> There are tens (hundreds?) of millions of mobile end-users who will be broken otherwise for IPv4-only sites.

I don't think anyone will be broken, because the alternative is to just ignore address hints you can't use.

> If we don't have a MUST on the client or the server...

We have a MUST on the client: "clients MUST perform A and/or AAAA queries for SvcDomainName as in {{client-behavior}}"

Reply:
Good point.  So perhaps:

When only "a" parameters are present, IPv6-only clients may either synthesize IPv6 addresses as specified in {{!RFC7050}} or fall back to resolving SvcDomainName. To avoid this, "aaaa" parameters SHOULD be published whenever "a" parameters are published.

?

Reply:
Works for me.

Reply:
OK, I added a version of this.  Note that "fallback" is probably not an accurate depiction of what's going on here: the AAAA query MUST be issued regardless, and SHOULD be used as soon as it's ready, so the only question is whether to wait for it.